### PR TITLE
 fixing save bug

### PR DESCRIPTION
**Pull request by nnmkhang**
[AB#34271](https://microsoftgarage.visualstudio.com/002806e2-ebaa-4672-9d2e-5fe5d29154ef/_workitems/edit/34271)

Please Re-Review. 

Overview: 
- Added SpinnerModal
- Added URLs for the designer 
- Changed ComponentDidMount logic in designer.
- Spinner now overlays when an action is in progress in the designer (ie Save) 

Snapshots ( Pls notice the urls ) :

Saved Card Before:
image

Saved Card After: 
image

New Card before:
image

New Card After:
image

Spinner:
image

Spinner after removing backgroud: 
image

### CODE REVIEW

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 1]
```diff
@@ -149,7 +149,7 @@ function initDesigner(): ACDesigner.CardDesigner {
 }
 
 function onSave(designer: ACDesigner.CardDesigner, props: DesignerProps): void {
-  if(props.templateID === ""){
+  if(props.templateID === "" || props.templateID === undefined){
```
Comment (beervoley) [inline private-templates-service/client/src/components/Designer/Designer.tsx]: Can substitute boolean expression to `if (not props.templateID)`

File: private-templates-service/client/src/components/NavBar/NavBar.tsx [hunk 2]
```diff
@@ -100,8 +100,8 @@ const NavBar = (props: NavBarProps) => {
         <Banner>
           <MobileBanner>
             <StyledLogo src={Logo} />
-            <Styledh1>{(props.templateID === "" && UNTITLEDCARD) || props.templateName}</Styledh1>
-            {props.templateID !== "" && <EditButton onClick={editName} iconProps={{ iconName: 'Edit' }} />}
+            <Styledh1>{((props.templateID === "" || props.templateID === undefined) && UNTITLEDCARD) || props.templateName}</Styledh1>
+            {(props.templateID !== "" && props.templateID !== undefined ) && <EditButton onClick={editName} iconProps={{ iconName: 'Edit' }} />}
```
Comment (beervoley) [inline private-templates-service/client/src/components/NavBar/NavBar.tsx]: Can also substitute boolean expression same as comment above

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    super(props);
+  }
+  
+  componentDidUpdate(){
+    if(!this.props.isFetching){
+      if(this.props.closeAction){
+        this.props.closeAction();
+      }
+      this.props.closeModal();
+    }
```
Comment (kodyang) [inline private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx]: Why does this spinner modal do actions passed in by parent components?

I think it makes sense for a spinner to be UI only, the parent component should handle side effects.

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 4]
```diff
@@ -60,18 +66,43 @@ interface State {
...
+  
+  updateURL = () => {
+      this.props.history.push("/designer/"+ this.props.templateID + "/" + this.props.version);
+  }
+  
+  toggleSpinner = (isFetching: boolean) => {
+    if(isFetching){
+      this.props.openModal(ModalState.Spinner)
+    }
+  }
```
Comment (kodyang) [inline private-templates-service/client/src/components/Designer/Designer.tsx]: Why do we do this:
`isFetching === true -> component did update -> dispatch action to open spinner modal -> ModalState === spinner -> spinner open`

Why cant we do this
`isFetching === true -> spinner open`
By extension, `isFetching === false -> spinner not open`

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 5]
```diff
@@ -122,9 +153,10 @@ class Designer extends React.Component<DesignerProps,State> {
     return (
       <React.Fragment>
         <DesignerWrapper id="designer-container" />
-        {this.props.modalState===ModalState.Save && <SaveModal designerSampleData = {designer.sampleData} designerTemplateJSON = {designer.getCard()}/>}
-        {this.props.modalState === ModalState.EditName && <EditNameModal />}
-      </React.Fragment>
+        {this.props.modalState === ModalState.Spinner && <SpinnerModal closeAction = {this.updateURL}/>}
```
Comment (kodyang) [inline private-templates-service/client/src/components/Designer/Designer.tsx]: could just make this 
`isFetching && <spinnermodal ...`

File: private-templates-service/client/src/components/NavBar/NavBar.tsx [hunk 6]
```diff
@@ -101,9 +101,9 @@ const NavBar = (props: NavBarProps) => {
       return (
         <Banner>
           <MobileBanner>
-            <StyledLogo aria-label={STRINGS.LOGO_DESCRIPTION} src={Logo} />
-            <Styledh1>{(props.templateID === "" && STRINGS.UNTITLEDCARD) || props.templateName}</Styledh1>
-            {props.templateID !== "" && <EditButton onClick={editName} iconProps={{ iconName: 'Edit' }} />}
+            <StyledLogo src={Logo} />
```
Comment (kodyang) [inline private-templates-service/client/src/components/NavBar/NavBar.tsx]: why did we get rid of aria label

**Comment by kodyang** [comment]
The spinner is hard to see, maybe have a small white square behind it

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx (hunk 3 above)
Comment (guptaharsh25) [reply private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx]: This was my idea. Do you have ideas for how to handle this logic in the parent component?

@nnmkhang maybe we do something like `prevProps.isFetching === true && this.props.isFetching === false` inside of `componentDidUpdate()` ? I think that might work, you can try it

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  constructor(props: Props) {
+    super(props);
+  }
+  
+  componentDidUpdate(){
+    if(!this.props.isFetching){
+      if(this.props.closeAction){
+        this.props.closeAction();
+      }
+      this.props.closeModal();
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx]: I agree with what Kody said in his comments, although we treat this component like a Modal it really isn't, it's just an overlay for when `isFetching === true`. I feel like the `openModal()` and `closeModal()` actions aren't needed in this case.

Comment (guptaharsh25) [reply private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx]: Also that implies we don't need the `ModalState` enum

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 8]
```diff
@@ -15,6 +15,8 @@ import { DesignerWrapper } from './styled';
 import EditNameModal from '../Common/EditNameModal';
 import SaveModal from './SaveModal/SaveModal';
 import { ModalState } from '../../store/page/types';
+import { withRouter, RouteComponentProps } from 'react-router-dom';
+import SpinnerModal from '../Common/SpinnerModal';
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/Designer.tsx]: Could you group similar imports together?

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 9]
```diff
@@ -60,18 +66,43 @@ interface State {
...
+  }
+  
+  toggleSpinner = (isFetching: boolean) => {
+    if(isFetching){
+      this.props.openModal(ModalState.Spinner)
+    }
+  }
+
+  componentDidUpdate() {
+    this.toggleSpinner(this.props.isFetching);
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/Designer.tsx]: won't this get triggered if ANY prop changes? i could be interpreting it wrong

File: private-templates-service/client/src/components/Designer/SaveModal/SaveModal.tsx [hunk 10]
```diff
@@ -87,7 +87,7 @@ class SaveModal extends React.Component<Props, State> {
       this.setState({ templateName: newValue });
     }
   }
-
+
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/SaveModal/SaveModal.tsx]: remove unnecessary whitespace?

File: private-templates-service/client/src/components/NavBar/NavBar.tsx [hunk 11]
```diff
@@ -101,9 +101,9 @@ const NavBar = (props: NavBarProps) => {
...
+            <Styledh1>{((props.templateID === "" || props.templateID === undefined) && STRINGS.UNTITLEDCARD) || props.templateName}</Styledh1>
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/NavBar/NavBar.tsx]: Correct me if I'm wrong but when the REST call is in progress I believe the template name becomes blank, can we handle that case using `isFetching`?

File: private-templates-service/client/src/components/SideBar/SideBar.tsx [hunk 12]
```diff
@@ -98,7 +98,7 @@ const navMenuLinks: INavLinkGroup[] = [
       },
       {
         name: "New Template",
-        url: "/designer",
+        url: "/designer/newcard/1.0",
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/SideBar/SideBar.tsx]: can we extract this into `const newTemplateURL = "/designer/newcard/1.0"` and reference that here and on line 158 in this file? Just so that we don't break anything by accident if this URL ever changes.

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx (hunk 3 above)
Comment (nnmkhang) [reply private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx:40]: Kody do you know a way to trigger the call on the falling edge of isFetching? that is what we are trying to do

File: private-templates-service/client/src/components/NavBar/NavBar.tsx (hunk 11 above)
Comment (nnmkhang) [reply private-templates-service/client/src/components/NavBar/NavBar.tsx:105]: I will create another ticket to address this thx

**Comment by nnmkhang** [comment]
> Quick idea that can be included in this PR: When the save modal opens can we populate the card name textfield with the current name of the template? I believe it'll always be "Untitled Card".
> 
> Not necessary but right now it looks like this:
> image
> 
> nav bar implies there is a name attached to this card but save modal doesn't reflect that

There is no opportunity to change the name for the first time when creating a new card until the user enters the save modal. That is why the name in the text field is always 'Untitled Card'

File: private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import styled from 'styled-components';
+import { Spinner } from 'office-ui-fabric-react/lib/Spinner';
+import { COLORS } from '../../../globalStyles';
+
+export const CenteredSpinner = styled(Spinner)`
+  margin: auto;
+`;
+
+export const Dot = styled.div`
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx:9]: What does this do? I saw that it wraps the spinner.

Comment (guptaharsh25) [reply private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx:9]: Is is just a white circular background that the spinner can sit on top off?

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 14]
```diff
@@ -60,18 +66,42 @@ interface State {
...
+
+  componentDidUpdate() {
+  if (this.props.location.pathname == '/designer/newcard/1.0' && this.props.templateID && this.props.version) {
+    this.props.history.replace('/designer/' + this.props.templateID + '/' + this.props.version);
+  }
+    this.toggleSpinner(this.props.isFetching);
+    if (this.props.templateJSON){
+      designer.setCard({...this.props.templateJSON});
+    }
+  }
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/Designer.tsx:104]: nit: formatting (whitespace) seems off here, but that might just be Github

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 15]
```diff
@@ -60,18 +66,42 @@ interface State {
...
-    this.state = { isSaveOpen: false };
+    this.state = {isSaveOpen: false };
+    if(this.props.match.params.uuid !== "newcard"){
+      this.props.getTemplate(this.props.match.params.uuid);
+    }
   }
 
   toggleModal = () => {
     this.setState({ isSaveOpen: !this.state.isSaveOpen });
   }
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/Designer.tsx:88]: is this used anywhere? I didn't see it. If not line 80 `this.state = {isSaveOpen: false };` and the State interface can be removed as well.

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 16]
```diff
@@ -99,7 +129,7 @@ class Designer extends React.Component<DesignerProps, State> {
     designer.monacoModuleLoaded(monaco);
 
     if (this.props.templateJSON) {
-      designer.setCard(this.props.templateJSON);
+      designer.setCard({...this.props.templateJSON});
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/Designer.tsx:132]: why did we have to do this again?

File: private-templates-service/client/src/components/Designer/SaveModal/SaveModal.tsx [hunk 17]
```diff
@@ -131,7 +131,7 @@ class SaveModal extends React.Component<Props, State> {
               </Container>
               <InfoWrapper>
                 <StyledH3>{CARDNAME}</StyledH3>
-                <StyledTextField onChange={this.onChange} />
+                <StyledTextField onChange={this.onChange} placeholder = {UNTITLEDCARD} />
```
Comment (guptaharsh25) [inline private-templates-service/client/src/components/Designer/SaveModal/SaveModal.tsx:134]: `placeholder` sets the greyed out unselectable text inside the textfield, `defaultValue` sets the initial value of the textfield.

IMO, `defaultValue={UNTITLEDCARD}` and `placeholder={"My Card"}`, something like that.

File: private-templates-service/client/src/store/page/types.ts [hunk 18]
```diff
@@ -11,7 +11,8 @@ export enum ModalState {
   Version,
   EditName,
   Delete,
-  Save
+  Save,
+  Spinner
```
Comment (guptaharsh25) [inline private-templates-service/client/src/store/page/types.ts:15]: I don't think you use this anymore.

File: private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx (hunk 13 above)
Comment (nnmkhang) [reply private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx:9]: yes

File: private-templates-service/client/src/components/Designer/Designer.tsx (hunk 16 above)
Comment (nnmkhang) [reply private-templates-service/client/src/components/Designer/Designer.tsx:132]: This is for when the designer is opened the second time. It has an issue with binding with designer.setCard(). This makes sure that we create a new object from the props and not use any weird binding from setCard()

**Comment by nnmkhang** [comment]
> good work comrade
Alt Text

File: private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  height: 30px;
+  width: 30px;
+  background-color: ${COLORS.WHITE};
+  border-radius: 50%;
```
Comment (kodyang) [inline private-templates-service/client/src/components/Common/SpinnerModal/styled.tsx:13]: nit
 ask naomi how the spinner background should look, I think its more standard to have a square background.

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+const mapStateToProps = (state: RootState) => {
+  return {
+    isFetching: state.currentTemplate.isFetching
+  }
+}
+
+interface Props {
+  isFetching: boolean;
+}
```
Comment (kodyang) [inline private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx:16]: Not used. remove

File: private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+      <BackDrop>
+        <Dot>
+          <CenteredSpinner size={SpinnerSize.large}/>
+        </Dot>
+      </BackDrop>
+    );
+  }
+}
+
+export default connect(mapStateToProps)(SpinnerModal);
```
Comment (kodyang) [inline private-templates-service/client/src/components/Common/SpinnerModal/SpinnerModal.tsx:34]: `export default SpinnerModal`

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 22]
```diff
@@ -38,39 +40,57 @@ const mapDispatchToProps = (dispatch: any) => {
...
-    this.state = { isSaveOpen: false };
+    if(this.props.match.params.uuid !== "newcard"){
+      this.props.getTemplate(this.props.match.params.uuid);
+    }
+  }
+
+  toggleSpinner = (isFetching: boolean) => {
+    if(isFetching){
+      this.props.openModal(ModalState.Spinner)
+    }
```
Comment (kodyang) [inline private-templates-service/client/src/components/Designer/Designer.tsx:83]: openModal no longer triggering spinner modal. delete

File: private-templates-service/client/src/components/Designer/Designer.tsx [hunk 23]
```diff
@@ -38,39 +40,57 @@ const mapDispatchToProps = (dispatch: any) => {
...
   }
 
-  toggleModal = () => {
-    this.setState({ isSaveOpen: !this.state.isSaveOpen });
+  componentDidUpdate() {
+    if (this.props.location.pathname == '/designer/newcard/1.0' && this.props.templateID && this.props.version) {
+      this.props.history.replace('/designer/' + this.props.templateID + '/' + this.props.version);
+    }
+    this.toggleSpinner(this.props.isFetching);
```
Comment (kodyang) [inline private-templates-service/client/src/components/Designer/Designer.tsx:90]: remove

File: private-templates-service/client/src/components/NavBar/NavBar.tsx [hunk 24]
```diff
@@ -101,11 +101,11 @@ const NavBar = (props: NavBarProps) => {
       return (
         <Banner>
           <MobileBanner>
-            <StyledLogo aria-label={STRINGS.LOGO_DESCRIPTION} src={Logo} />
-            <Styledh1>{(props.templateID === "" && STRINGS.UNTITLEDCARD) || props.templateName}</Styledh1>
-            {props.templateID !== "" && <EditButton onClick={editName} iconProps={{ iconName: 'Edit' }} />}
+          <StyledLogo aria-label={STRINGS.LOGO_DESCRIPTION} src={Logo} />
```
Comment (kodyang) [inline private-templates-service/client/src/components/NavBar/NavBar.tsx:104]: nit: tab
